infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/KubernetesPluginsToolingApplierTest.java

```diff
@@ -102,11 +107,137 @@ public void setUp() {
+  public void shouldProvisionPluginsCommandsToEnvironment() throws Exception {
+    // given
+    Command pluginCommand =
+        new Command()
+            .name("test-command")
+            .workingDir("~")
+            .command(Arrays.asList("./build.sh", "--no-pull"));
+
+    // when
+    applier.apply(
+        internalEnvironment,
+        singletonList(createChePlugin("plugin", createContainer("container", pluginCommand))));
+
+    // then
+    List<CommandImpl> envCommands = internalEnvironment.getCommands();
+    assertEquals(envCommands.size(), 1);
+    CommandImpl envCommand = envCommands.get(0);
+    assertEquals(envCommand.getName(), pluginCommand.getName());
+    assertEquals(
+        envCommand.getCommandLine(),
+        pluginCommand.getCommand().stream().collect(Collectors.joining(" ")));
+    assertEquals(envCommand.getType(), "custom");
+    assertEquals(envCommand.getAttributes().get("workDir"), pluginCommand.getWorkingDir());
+    assertEquals(envCommand.getAttributes().get("machineName"), POD_NAME + "/plugin-container");
+  }
+
+  @Test
+  public void shouldResolveMachineNameForCommandsInEnvironment() throws Exception {
+    // given
+    ChePlugin chePlugin = createChePlugin("plugin", createContainer("container"));
+    String pluginRef = chePlugin.getId() + ":" + chePlugin.getVersion();
+
+    CommandImpl pluginCommand = new CommandImpl("test-command", "echo Hello World!", "custom");
+    pluginCommand.getAttributes().put("plugin", pluginRef);
+    internalEnvironment.getCommands().add(pluginCommand);
+
+    // when
+    applier.apply(internalEnvironment, singletonList(chePlugin));
+
+    // then
+    List<CommandImpl> envCommands = internalEnvironment.getCommands();
+    assertEquals(envCommands.size(), 1);
+    CommandImpl envCommand = envCommands.get(0);
+    assertEquals(envCommand.getName(), pluginCommand.getName());
+    assertEquals(envCommand.getType(), pluginCommand.getType());
+    assertEquals(envCommand.getCommandLine(), pluginCommand.getCommandLine());
+    assertEquals(envCommand.getAttributes().get("plugin"), pluginRef);
+    assertEquals(envCommand.getAttributes().get("machineName"), POD_NAME + "/plugin-container");
+  }
+
+  @Test
+  public void shouldFillInWarningIfChePluginDoesNotHaveAnyContainersButThereAreRelatedCommands()
+      throws Exception {
+    // given
+    ChePlugin chePlugin = createChePlugin("plugin");
+    String pluginRef = chePlugin.getId() + ":" + chePlugin.getVersion();
+
+    CommandImpl pluginCommand = new CommandImpl("test-command", "echo Hello World!", "custom");
+    pluginCommand.getAttributes().put("plugin", pluginRef);
+    internalEnvironment.getCommands().add(pluginCommand);
+
+    // when
+    applier.apply(internalEnvironment, singletonList(chePlugin));
+
+    // then
+    List<Warning> envWarnings = internalEnvironment.getWarnings();
+    assertEquals(envWarnings.size(), 1);
+    Warning warning = envWarnings.get(0);
+    assertEquals(warning.getCode(), 44012);
```

Comment:
It is better to reuse constants rather than use magic numbers

Reply:
Agree, I'll find a suitable place for warnings codes constant.

Reply:
Fixed 
in https://github.com/eclipse/che/pull/12016/commits/4f45ebbd7618b17f9e9687961e852cd47069ad3e